azure/services/async/async.go

```diff
@@ -149,9 +162,11 @@ func (s *Service) DeleteResource(ctx context.Context, spec azure.ResourceSpecGet
 
 	// Check if there is an ongoing long running operation.
 	future := s.Scope.GetLongRunningOperationState(resourceName, serviceName)
-	if future != nil {
-		_, err := processOngoingOperation(ctx, s.Scope, s.Deleter, resourceName, serviceName)
-		return err
+	// PUT operations still in progress should not be waited on if a delete was requested.
+	if future != nil && future.Type == infrav1.DeleteFuture {
```

Comment:
we also want the reverse to be true (don't block a PUT on a delete operation in progress). What do you think of modifying `GetLongRunningOperationState` to only return futures for the right operation type so this applies to all operations?

Reply:
That makes sense, I'll add that.

I noticed there are a couple of places this is called outside of this file for scalesets:

Since it looks like this is looking for a future of any type, is it worth baking in some capability to search for multiple types? Or maybe just do that inline here?
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d28ff9a7d44b25fe10ca3e13a7e2725b50b35fad/azure/services/scalesets/scalesets.go#L94

The second one is in a `Delete` method, so I'm guessing we'd only want delete futures here too?
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d28ff9a7d44b25fe10ca3e13a7e2725b50b35fad/azure/services/scalesets/scalesets.go#L185

Reply:
The first use (L94) is in Reconcile so that's either a PUT or PATCH future, and the second use (L185) should be just DELETE futures.

Scalesets still needs to be refactored as part of #1702, after that it will also be using async.go, so I think we can just do it inline for now and check both.